proto/src/rr/dnssec/rdata/dnskey.rs

```diff
@@ -347,8 +347,13 @@ pub fn read(decoder: &mut BinDecoder, rdata_length: u16) -> ProtoResult<DNSKEY>
     let algorithm: Algorithm = Algorithm::read(decoder)?;
 
     // the public key is the left-over bytes minus 4 for the first fields
-    // TODO: decode the key here?
-    let public_key: Vec<u8> = decoder.read_vec((rdata_length - 4) as usize)?;
+    //   this sub is safe, as the first 4 fields must have been in the rdata, otherwise there would have been
+    //   an earlier return.
+    let key_len = rdata_length
+        .map(|u| u as usize)
+        .checked_sub(4)
```

Comment:
FYI @oherrala here's another example of the issues you raised, by packaging the rdata_length in Restrict it makes these situations more aparent.